Header block

davidpmccormick opened the pull request:
This adds styles and markup for article title blocks and library item title blocks ~and discover landing block items~.

__TODO:__
- [ ] Handle overflow display of long titles (library item specific).

![screen shot 2016-11-17 at 16 10 33](https://cloud.githubusercontent.com/assets/1394592/20397441/a630bb7e-ace1-11e6-8557-7b33523a244b.png)
![screen shot 2016-11-17 at 16 10 44](https://cloud.githubusercontent.com/assets/1394592/20397444/a63148f0-ace1-11e6-8f63-0c9abe4ce1cb.png)

jamesgorrie commented on server/views/components/header-block/index.njk:
```diff
@@ -0,0 +1,10 @@
+<header class="header-block {% for modifier in modifiers %}header-block--{{ modifier }} {% endfor %}">
```
Again, composing CSS names is almost anti-BEM - and makes static analysis hard.
Could we not do something easier and just have extra classes which is more helpful for practical use?

jamesgorrie replied:
And it's used elsewhere.

jamesgorrie commented on client/scss/components/_header_block.scss:
```diff
@@ -0,0 +1,107 @@
+      position: relative;
+      padding: 1.8rem 0 1.5rem span-columns('large', 1.5);
+
+      &:before {
```
Could this just be a border?

davidpmccormick replied:
Border widths can't be (or include) percentages, so sadly not.

jamesgorrie commented on client/scss/components/_header_block.scss:
```diff
@@ -0,0 +1,107 @@
+  margin-top: 0;
+  margin-left: -0.06em;
+
+  .header-block--a & {
```
This is starting to break BEM again.
Could we not set the state on the title?

i.e.
```
.header-block__title--a {}
```

Also - what is `--a`?

davidpmccormick replied:
I went with `a`, `b` and `c` after chucking in the idea of calling them `article`, `library-item` and `discover-landing`. Better to keep the naming more abstract than just the _original_ purpose for which they were designed.

Happy to take other suggestions (I went via, `alpha`, `beta` and `gamma`, but thought that that was ultimately more unnecessary cruft. And my Greek alphabet knowledge gets a bit hazy thereafter).

jamesgorrie replied:
Could it not just explain a little more what it's doing with the styling? I still cant' tell.

I think we should be using SASS for styling, and not, like you said, their purpose.

jamesgorrie commented on client/scss/components/_header_block.scss:
```diff
@@ -0,0 +1,107 @@
+    @include font('HNM1');
+  }
+
+  .header-block--c.header-block--small & {
```
Composing classes like this is defeating of a lot of the BEM principles. It means the SASS is composed, where as we want to be able to do this in the HTML.

Adding more classes is more BEMy.

We could have:
`.header-block--c { @include font('WB6'); }`
`.header-block--small { margin-bottom: 0.4em; }`
```

Then I can add those classes wherever I want.

davidpmccormick replied:
I know what you mean. But then I also think what's nice about having the sass componentized is that you can reason comfortably about an element within that component and not have to worry about specificity wars.

I've previously found using a parent selector to scope modified elements within it has been a decent compromise, but I'm willing to try a firmer approach if that's what we want to do.

davidpmccormick commented:
This component was working too hard – I've removed the `--c` variant ('discover landing block') leaving the two page-description components (currently designed for articles and library items). This makes reasoning about styles easier. Renamed to `page-description` accordingly.

Variant stylings are now written long-hand (for ease of static analysis) and applied directly to the html that they modify (to be more BEM).

Will re-instate the removed `--c` variant as a `block-description` in a new PR.

jamesgorrie commented:
Maybe I am not understanding properly - but what do the modifiers `--wb` and `--hn` mean?

davidpmccormick commented:
> what do the modifiers --wb and --hn mean?

'wellcome bold' and 'helvetica neue' (the font family of the heading)

jamesgorrie commented:
It feels as though we're sitting in between two methodologies as I can see it.

We could move down the route of composing classes on the `HTML` element. e.g:
```SaSS
.page-description { /*...*/ }
/*...*/
.ff--hn { font-face: Helvetica Neue; }
```
```HTML
<h1 class="page-description ff--hn">Heading</h1>
```

Or we could go about composing this in the SASS:
```SaSS
.page-description {/*...*/}
.page-description--article { @extend .ff--hn }
/*...*/
.ff--hn { font-face: Helvetica Neue; }
```
```HTML
<h1 class="page-description page-description--article">Heading</h1>
```


In version 1 we leave styling and classes to do exactly that.
In 2 we add some semantics, offering what's special about this element and then decide to style it.

I personally don't like 2 as it couples semantics and style.

What we seem to have at the moment is a kind-of version of 1, but also coupled with other styles attached to it - and I can't see the reason why?

davidpmccormick commented:
I agree – I don't think I like `--article` because it's a bit too prescriptive (we might want to use this title style on something other than an article in the future, even thought that's what it was originally designed for). So the modifier differential was to distinguish between the two based on the obvious difference between them stylistically (rather than `--a` and `--b`, as previously discussed).

This assumes that we're going to treat this `page-description` as a type of component which has a couple of variants. This is the approach that I had understood we were going to take – we split the designs up into bits that we agree make sense to share the same markup and styles with variants and code them up.

I would argue that having a class per font-style goes too far – I'd prefer that a e.g. page-description is responsible for its own styles.

jamesgorrie commented:
> I would argue that having a class per font-style goes too far – I'd prefer that a e.g. page-description is responsible for its own styles.

It might make more sense to have a different modifier then - not one that does more than it says on the tin - do we know what these states are going to be - or are we hypotheticalising about them?

davidpmccormick commented:
Currently we know that they've been designed to handle the article and library item title blocks – but all of the naming issues we've raised before still stand (i.e. don't want to call them `--article` and `--library` because we might want to use them elsewhere; don't want to call them `--a` and `--b` because it doesn't describe styles).

jamesgorrie commented:
So article and library are being handled differently?
I am assuming the reasoning for this is the idea behind metadata vs editorial content.

We could then have a `--metadata` modifier that turns it into Helvetica?

jamesgorrie commented:
Although taking it a step further - we could completely decouple it and have:
```
.content-heading {}
.metadata-heading {}
```
?

davidpmccormick commented:
I believe the difference is simply to set them apart, visually. One style for articles, another for library items. We're in danger of overthinking it.

jamesgorrie commented:
> it has the least potential to become out-of-date

My concern is not that - my concern is that it is doing more than that - there is padding etc being set too.

> We're in danger of overthinking it

My problem thus far is I am still unclear what is going on - so I'm looking for clarity, not extra information. Let me think of how to explain my unclearness more clearly...

jamesgorrie commented on client/scss/components/_page_description.scss:
```diff
@@ -0,0 +1,66 @@
+//* @define page-description;
+
+.page-description {
+  &.page-description--wb {
```
So here this is Wellcome Bold - but nowhere in the styling is that applied.
There is also a lot of extra styling that has nothing to do with the font.

davidpmccormick replied:
yeah, you're right – in searching for what set the two variants apart the most, it was the font-family that was the winner, but this isn't ideal, granted.

jamesgorrie commented on client/scss/components/_page_description.scss:
```diff
@@ -0,0 +1,66 @@
+
+  &.page-description--hn {
+    padding-bottom: 1.2rem;
+    border-bottom: 1px solid color('keyline-grey');
```
This hasn't actually set the style to Helvetica?

davidpmccormick replied:
yep you're right, same as above.

jamesgorrie commented on client/scss/components/_page_description.scss:
```diff
@@ -0,0 +1,66 @@
+
+  &.page-description__title--wb {
+    @include font('WB3');
+    margin-bottom: 0.25em;
```
Why does WB get this - if it's a universal when applying WB could we wrap it in a font-include mixin to make sure we don't omit it later?

davidpmccormick replied:
This defines the text lock-up within the component – I don't think it can be abstracted away to universal font styling.

jamesgorrie commented on client/scss/components/_page_description.scss:
```diff
@@ -0,0 +1,66 @@
+
+  &.page-description__outro--wb {
+    @include font('HNL5');
+  }
```
Why don't we have WB modifiers here?

jamesgorrie commented on server/views/components/page-description/index.njk:
```diff
@@ -0,0 +1,10 @@
+<header class="page-description {% for modifier in modifiers %}page-description--{{ modifier }} {% endfor %}">
+  <span class="page-description__intro {% for modifier in modifiers %}page-description__intro--{{ modifier }} {% endfor %}">{{ intro }}</span>
+  <h1 class="page-description__title {% for modifier in modifiers %}page-description__title--{{ modifier }} {% endfor %}">{{ title }}</h1>
+  <span class="page-description__outro {% for modifier in modifiers %}page-description__outro--{{ modifier }} {% endfor %}">
```
So we'll have the `--wb` modifier written, but nothing will happen?

davidpmccormick replied:
Yeah, this is rubbish isn't it? I think in an attempt to cover off all your points, I'm increasingly unhappy with what's going on and I'm wasting your time a bit.

I'm more in favour of going back to the component-scoped style that I started out with. 

You mentioned above that we had decided not to go down that route, but I wasn't aware of that – I was very much under the impression that we were going to take a component driven approach.

Would you object to me having another bash at it that way round?

jennpb commented:
To add some clarity, the reason for the WB and HN difference between library content and everything else is because library content (specifically, items from the collections) contain titles that are 1) too long to be legible in WB and 2) may contain glyphs and diacritics not present in WB. 'Embellished' and 'plain' seem meaningful to me fwiw

jamesgorrie commented:
@jennpb if that's the case - I would prefer to come up with a css name that conveys that.
`heading--short`, `heading--long` @davidpmccormick?

davidpmccormick commented:
The trouble as I see it is I think e.g. `--short`/`--long` or `--embellished`/`--plain` run the risk of painting us into a corner really quickly (semantics-wise). Keeping it more abstract (`--a`/`--b`) doesn't create any problems as far as I'm aware? And when there's a third style of heading in the future that is a _bit_ embellished, or longer-but-plainer, the more specific naming system breaks really quickly.
